hazelcast/src/main/java/com/hazelcast/client/impl/spi/ProxyManager.java

```diff
@@ -230,34 +283,39 @@ private ClientProxy getOrCreateProxyInternal(@Nonnull String service,
         checkNotNull(service, "Service name is required!");
         checkNotNull(id, "Object name is required!");
 
-        final ObjectNamespace ns = new DistributedObjectNamespace(service, id);
-        ClientProxyFuture proxyFuture = proxies.get(ns);
-        if (proxyFuture != null) {
-            return proxyFuture.get();
-        }
-        ClientProxyFactory factory = proxyFactories.get(service);
-        if (factory == null) {
-            throw new ClientServiceNotFoundException("No factory registered for service: " + service);
-        }
-        proxyFuture = new ClientProxyFuture();
-        ClientProxyFuture current = proxies.putIfAbsent(ns, proxyFuture);
-        if (current != null) {
-            return current.get();
-        }
-
+        distributedObjectsReadLock.lock();
         try {
-            ClientProxy clientProxy = createClientProxy(id, factory);
-            if (remote) {
-                initialize(clientProxy);
-            } else {
-                clientProxy.onInitialize();
+            final ObjectNamespace ns = new DistributedObjectNamespace(service, id);
+            ClientProxyFuture proxyFuture = proxies.get(ns);
+            if (proxyFuture != null) {
+                return proxyFuture.get();
             }
-            proxyFuture.set(clientProxy);
-            return clientProxy;
-        } catch (Throwable e) {
-            proxies.remove(ns);
-            proxyFuture.set(e);
-            throw rethrow(e);
+            ClientProxyFactory factory = proxyFactories.get(service);
+            if (factory == null) {
+                throw new ClientServiceNotFoundException("No factory registered for service: " + service);
+            }
+            proxyFuture = new ClientProxyFuture();
+            ClientProxyFuture current = proxies.putIfAbsent(ns, proxyFuture);
```

Comment:
you are doing a modification here, i.e. a `write` operation. But you only have the read lock. I would expect you to upgrade to write lock at this point before you make the change.

Reply:
I used a write lock after some point.

Reply:
I realized that I should not use a write lock here. The important thing is not doing a write operation or read operation. In a readwriteLock readers can lock read lock multiple times and write lock can only be locked once. What I want to achieve is that if no polling in progress, multiple proxy creation in parallel should be possible, thus use a read lock. And for polling distributed objects, use a write lock so that no proxy creation is possible during that. 

I am worried about the following case: what happens if a lot of proxies are created in parallel continously ? In this case polling distributed objects task won't be able to run. I think this is related to fairness of the readwriteLock implementation. I wonder you opinion too, @sancar, what do you think about this scenario?

Reply:
Not a common scenario. Ye, you may not get up-to-date data for some time but how long can this continue. It is not a usable system anymore if that happens. I think the implementation is good enough.